File: Android/app/src/main/java/com/example/supplementtracker/service/SyncOperationJournal.kt
package com.example.supplementtracker.service

import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject

enum class SyncJournalEvent {
    START,
    CONFLICT,
    BACKOFF,
    SUCCESS,
    FAILURE
}

data class SyncConflictPreview(
    val remoteWins: Int = 0,
    val localWins: Int = 0,
    val tieLocalWins: Int = 0
) {
    val materialConflictCount: Int get() = remoteWins + localWins + tieLocalWins
}

enum class SyncConflictDecision {
    REMOTE_WINS,
    LOCAL_WINS,
    TIE_LOCAL_WINS,
    NO_CONFLICT
}

object SyncConflictPolicy {
    fun decide(localExists: Boolean, localTs: Long, remoteTs: Long, differs: Boolean): SyncConflictDecision {
        if (!localExists || !differs) return SyncConflictDecision.NO_CONFLICT
        if (remoteTs > localTs) return SyncConflictDecision.REMOTE_WINS
        if (remoteTs < localTs) return SyncConflictDecision.LOCAL_WINS
        return SyncConflictDecision.TIE_LOCAL_WINS
    }

    fun remoteMayApply(localTs: Long, remoteTs: Long): Boolean = remoteTs > localTs
}

data class SyncRetryState(
    val failureCount: Int,
    val nextRetryEpochMs: Long
)

object SyncBackoffPolicy {
    private const val BASE_DELAY_MS = 15_000L
    private const val MAX_DELAY_MS = 10 * 60_000L

    fun nextState(previousFailures: Int, nowEpochMs: Long): SyncRetryState {
        val failures = (previousFailures + 1).coerceAtMost(16)
        return SyncRetryState(failures, nowEpochMs + delayMs(failures))
    }

    fun delayMs(failureCount: Int): Long {
        if (failureCount <= 0) return 0L
        val exponent = (failureCount - 1).coerceAtMost(6)
        return (BASE_DELAY_MS * (1L shl exponent)).coerceAtMost(MAX_DELAY_MS)
    }

    fun canAttempt(state: SyncRetryState, nowEpochMs: Long): Boolean =
        state.nextRetryEpochMs <= 0L || nowEpochMs >= state.nextRetryEpochMs
}

class SyncRetryStore(private val prefs: SharedPreferences) {
    fun state(manifestId: String): SyncRetryState = SyncRetryState(
        prefs.getInt(failureKey(manifestId), 0),
        prefs.getLong(nextRetryKey(manifestId), 0L)
    )

    fun recordFailure(manifestId: String, nowEpochMs: Long = System.currentTimeMillis()): SyncRetryState {
        val next = SyncBackoffPolicy.nextState(state(manifestId).failureCount, nowEpochMs)
        prefs.edit().putInt(failureKey(manifestId), next.failureCount)
            .putLong(nextRetryKey(manifestId), next.nextRetryEpochMs).apply()
        return next
    }

    fun clear(manifestId: String) {
        prefs.edit().remove(failureKey(manifestId)).remove(nextRetryKey(manifestId)).apply()
    }

    private fun failureKey(id: String): String = "cloudSyncRetryFailureCount_${id.trim()}"
    private fun nextRetryKey(id: String): String = "cloudSyncNextRetryEpochMs_${id.trim()}"
}

data class SyncJournalEntry(
    val epochMs: Long,
    val event: SyncJournalEvent,
    val stackDirty: Boolean = false,
    val historyDirty: Boolean = false,
    val retryCount: Int = 0,
    val nextRetryEpochMs: Long = 0L,
    val remoteWins: Int = 0,
    val localWins: Int = 0,
    val tieLocalWins: Int = 0
)

class SyncOperationJournalStore(private val prefs: SharedPreferences) {
    fun append(manifestId: String, entry: SyncJournalEntry) {
        val id = manifestId.trim()
        if (id.isEmpty()) return
        val array = readArray(id)
        array.put(encode(entry))
        prefs.edit().putString(key(id), trim(array).toString()).apply()
    }

    fun count(manifestId: String): Int = readArray(manifestId.trim()).length()

    private fun readArray(id: String): JSONArray {
        val raw = prefs.getString(key(id), null)
        return runCatching { if (raw.isNullOrBlank()) JSONArray() else JSONArray(raw) }
            .getOrElse { JSONArray() }
    }

    private fun encode(entry: SyncJournalEntry): JSONObject = JSONObject()
        .put("ts", entry.epochMs)
        .put("event", entry.event.name)
        .put("stackDirty", entry.stackDirty)
        .put("historyDirty", entry.historyDirty)
        .put("retryCount", entry.retryCount)
        .put("nextRetryEpochMs", entry.nextRetryEpochMs)
        .put("remoteWins", entry.remoteWins)
        .put("localWins", entry.localWins)
        .put("tieLocalWins", entry.tieLocalWins)

    private fun trim(array: JSONArray): JSONArray {
        val result = JSONArray()
        val start = (array.length() - 40).coerceAtLeast(0)
        for (index in start until array.length()) result.put(array.getJSONObject(index))
        return result
    }

    private fun key(id: String): String = "cloudSyncJournal_$id"
}

File: iOS/Services/SyncOperationJournal.swift
import Foundation

enum SyncMutationPart: String, Codable, CaseIterable, Hashable, Sendable {
    case stack
    case history
}

struct SyncMutationEntry: Codable, Equatable, Sendable {
    let part: SyncMutationPart
    let enqueuedAtEpochMs: Int64
}

enum SyncMutationQueueStore {
    static func markDirty(
        clientId: UUID,
        part: SyncMutationPart,
        nowEpochMs: Int64,
        defaults: UserDefaults = .standard
    ) {
        var entries = Dictionary(uniqueKeysWithValues: pending(clientId: clientId, defaults: defaults).map { ($0.part, $0) })
        let current = entries[part]?.enqueuedAtEpochMs ?? 0
        entries[part] = SyncMutationEntry(part: part, enqueuedAtEpochMs: max(current, nowEpochMs))
        save(Array(entries.values), clientId: clientId, defaults: defaults)
    }

    static func pending(clientId: UUID, defaults: UserDefaults = .standard) -> [SyncMutationEntry] {
        guard let data = defaults.data(forKey: key(clientId)) else { return [] }
        return ((try? JSONDecoder().decode([SyncMutationEntry].self, from: data)) ?? [])
            .sorted { $0.part.rawValue < $1.part.rawValue }
    }

    static func clearSynced(
        clientId: UUID,
        parts: Set<SyncMutationPart>,
        syncStartedEpochMs: Int64,
        defaults: UserDefaults = .standard
    ) {
        let remaining = pending(clientId: clientId, defaults: defaults).filter {
            !parts.contains($0.part) || $0.enqueuedAtEpochMs > syncStartedEpochMs
        }
        save(remaining, clientId: clientId, defaults: defaults)
    }

    private static func save(_ entries: [SyncMutationEntry], clientId: UUID, defaults: UserDefaults) {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        defaults.set(data, forKey: key(clientId))
    }

    private static func key(_ clientId: UUID) -> String {
        "cloudSyncMutationQueue_\(clientId.uuidString.lowercased())"
    }
}

enum SyncConflictDecision: String, Equatable, Sendable {
    case remoteWins
    case localWins
    case tieLocalWins
    case noConflict
}

struct SyncConflictPreview: Equatable, Sendable {
    var remoteWins: Int = 0
    var localWins: Int = 0
    var tieLocalWins: Int = 0

    var materialConflictCount: Int { remoteWins + localWins + tieLocalWins }
}

enum SyncConflictPolicy {
    static func decide(localExists: Bool, localTs: Int64, remoteTs: Int64, differs: Bool) -> SyncConflictDecision {
        guard localExists, differs else { return .noConflict }
        if remoteTs > localTs { return .remoteWins }
        if remoteTs < localTs { return .localWins }
        return .tieLocalWins
    }

    static func remoteMayApply(localTs: Int64, remoteTs: Int64) -> Bool {
        remoteTs > localTs
    }
}

struct SyncRetryState: Equatable, Sendable {
    let failureCount: Int
    let nextRetryEpochMs: Int64
}

enum SyncBackoffPolicy {
    private static let baseDelayMs: Int64 = 15_000
    private static let maxDelayMs: Int64 = 10 * 60_000

    static func nextState(previousFailures: Int, nowEpochMs: Int64) -> SyncRetryState {
        let failures = min(max(previousFailures + 1, 1), 16)
        return SyncRetryState(
            failureCount: failures,
            nextRetryEpochMs: nowEpochMs + delayMs(failureCount: failures)
        )
    }

    static func delayMs(failureCount: Int) -> Int64 {
        guard failureCount > 0 else { return 0 }
        let exponent = min(failureCount - 1, 6)
        let delay = baseDelayMs * Int64(1 << exponent)
        return min(delay, maxDelayMs)
    }

    static func canAttempt(_ state: SyncRetryState, nowEpochMs: Int64) -> Bool {
        state.nextRetryEpochMs <= 0 || nowEpochMs >= state.nextRetryEpochMs
    }
}

enum SyncRetryStore {
    static func state(manifestId: String, defaults: UserDefaults = .standard) -> SyncRetryState {
        SyncRetryState(
            failureCount: defaults.integer(forKey: failureKey(manifestId)),
            nextRetryEpochMs: Int64(defaults.double(forKey: nextRetryKey(manifestId)))
        )
    }

    static func recordFailure(
        manifestId: String,
        nowEpochMs: Int64,
        defaults: UserDefaults = .standard
    ) -> SyncRetryState {
        let next = SyncBackoffPolicy.nextState(
            previousFailures: state(manifestId: manifestId, defaults: defaults).failureCount,
            nowEpochMs: nowEpochMs
        )
        defaults.set(next.failureCount, forKey: failureKey(manifestId))
        defaults.set(Double(next.nextRetryEpochMs), forKey: nextRetryKey(manifestId))
        return next
    }

    static func clear(manifestId: String, defaults: UserDefaults = .standard) {
        defaults.removeObject(forKey: failureKey(manifestId))
        defaults.removeObject(forKey: nextRetryKey(manifestId))
    }

    private static func failureKey(_ id: String) -> String { "cloudSyncRetryFailureCount_\(id)" }
    private static func nextRetryKey(_ id: String) -> String { "cloudSyncNextRetryEpochMs_\(id)" }
}

enum SyncJournalEvent: String, Codable, Sendable {
    case start
    case conflict
    case backoff
    case success
    case failure
}

struct SyncJournalEntry: Codable, Equatable, Sendable {
    let epochMs: Int64
    let event: SyncJournalEvent
    var stackDirty: Bool = false
    var historyDirty: Bool = false
    var retryCount: Int = 0
    var nextRetryEpochMs: Int64 = 0
    var remoteWins: Int = 0
    var localWins: Int = 0
    var tieLocalWins: Int = 0
}

enum SyncOperationJournalStore {
    static func append(
        manifestId: String,
        entry: SyncJournalEntry,
        defaults: UserDefaults = .standard
    ) {
        let id = manifestId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !id.isEmpty else { return }
        var entries = read(manifestId: id, defaults: defaults)
        entries.append(entry)
        if entries.count > 40 { entries = Array(entries.suffix(40)) }
        guard let data = try? JSONEncoder().encode(entries) else { return }
        defaults.set(data, forKey: key(id))
    }

    static func count(manifestId: String, defaults: UserDefaults = .standard) -> Int {
        read(manifestId: manifestId, defaults: defaults).count
    }

    private static func read(manifestId: String, defaults: UserDefaults) -> [SyncJournalEntry] {
        guard let data = defaults.data(forKey: key(manifestId)) else { return [] }
        return (try? JSONDecoder().decode([SyncJournalEntry].self, from: data)) ?? []
    }

    private static func key(_ id: String) -> String { "cloudSyncJournal_\(id)" }
}
